File: src/services/voyageai_service.py
import os
import voyageai
from PIL import Image
from io import BytesIO
from typing import List, Union
from utils.config import VOYAGEAI_API_KEY
from utils.logger import logger
from utils.image_utils import compress_image, validate_image_for_voyage, get_image_info
from utils.singleton import Singleton


class VoyageAiService(metaclass=Singleton):
  def __init__(self, model="voyage-multimodal-3"):
    self.vo = voyageai.Client(VOYAGEAI_API_KEY)
    self.model = model

  def generate_text_embeddings(self, text: str):
    """Generate embeddings for a single text."""
    result = self.vo.multimodal_embed([[text]], model=self.model)
    return result.embeddings[0]

  def generate_image_embeddings(self, img_bytes: Union[bytes, BytesIO]) -> List[float]:
    """
    Generate embeddings for a single image.
    Automatically compresses image if needed to meet Voyage AI requirements.
    """
    # Convert BytesIO to bytes if needed
    if isinstance(img_bytes, BytesIO):
      img_bytes = img_bytes.getvalue()

    # Log original image info
    img_info = get_image_info(img_bytes)
    logger.info(
      f"Original image: {img_info.get('width')}x{img_info.get('height')} "
      f"({img_info.get('pixels', 0):,} pixels, {img_info.get('size_mb', 0):.2f}MB)"
    )

    # Check if compression is needed
    is_valid, error_msg = validate_image_for_voyage(img_bytes)

    if not is_valid:
      logger.warning(f"Image validation failed: {error_msg}. Compressing...")
      img_bytes = compress_image(img_bytes)

      # Validate after compression
      is_valid, error_msg = validate_image_for_voyage(img_bytes)
      if not is_valid:
        raise ValueError(f"Image still invalid after compression: {error_msg}")

      logger.info("Image compressed successfully and validated")
    else:
      logger.info("Image meets requirements, no compression needed")

    # Generate embeddings
    img = Image.open(BytesIO(img_bytes))
    result = self.vo.multimodal_embed([[img]], model=self.model)
    return result.embeddings[0]

  def generate_batch_embeddings(
    self, inputs: List[Union[str, bytes, Image.Image, BytesIO]]
  ) -> List[List[float]]:
    """
    Generate embeddings for multiple inputs (text and/or images) in one API call.
    Automatically compresses images if needed.

    Args:
        inputs: List of strings (text) or bytes/PIL Images (images)

    Returns:
        List of embeddings, one per input
    """
    prepared_inputs = []

    for idx, item in enumerate(inputs):
      if isinstance(item, str):
        # Text input
        prepared_inputs.append([item])
      elif isinstance(item, (bytes, BytesIO)):
        # Image bytes - compress if needed
        img_bytes = item.getvalue() if isinstance(item, BytesIO) else item

        # Log and validate
        img_info = get_image_info(img_bytes)
        logger.info(
          f"Input {idx}: Image {img_info.get('width')}x{img_info.get('height')} "
          f"({img_info.get('pixels', 0):,} pixels, {img_info.get('size_mb', 0):.2f}MB)"
        )

        is_valid, error_msg = validate_image_for_voyage(img_bytes)

        if not is_valid:
          logger.warning(f"Input {idx}: Compression needed - {error_msg}")
          img_bytes = compress_image(img_bytes)

          # Validate after compression
          is_valid, error_msg = validate_image_for_voyage(img_bytes)
          if not is_valid:
            raise ValueError(
              f"Input {idx}: Still invalid after compression: {error_msg}"
            )

          logger.info(f"Input {idx}: Compressed successfully")
        else:
          logger.info(f"Input {idx}: No compression needed")

        img = Image.open(BytesIO(img_bytes))
        prepared_inputs.append([img])
      elif isinstance(item, Image.Image):
        # Already a PIL Image - check if we need to compress
        output = BytesIO()
        item.save(output, format="JPEG", quality=85)
        img_bytes = output.getvalue()

        is_valid, error_msg = validate_image_for_voyage(img_bytes)

        if not is_valid:
          logger.warning(f"Input {idx}: PIL Image needs compression - {error_msg}")
          img_bytes = compress_image(img_bytes)
          img = Image.open(BytesIO(img_bytes))
        else:
          img = item

        prepared_inputs.append([img])
      else:
        raise ValueError(f"Unsupported input type at index {idx}: {type(item)}")

    # Single API call for all inputs
    logger.info(f"Generating embeddings for {len(prepared_inputs)} inputs...")
    result = self.vo.multimodal_embed(prepared_inputs, model=self.model)
    logger.info(f"Successfully generated {len(result.embeddings)} embeddings")

    return result.embeddings

File: src_go/internal/services/voyageai/voyageai_service.go
package voyageai

import (
	"bytes"
	"encoding/json"
	"fmt"
	_ "image/gif"
	_ "image/png"
	"io"
	"log/slog"
	"net/http"
	"os"
	"time"

	_ "golang.org/x/image/webp"
)

// SVC is the global VoyageAI service instance
var SVC *VoyageAIService

// Setup initializes the global VoyageAI service
func Setup() {
	SVC = &VoyageAIService{
		apiKey: os.Getenv("VOYAGEAI_API_KEY"),
		model:  defaultModel,
		client: &http.Client{
			Timeout: 60 * time.Second,
		},
	}
}

// GenerateTextEmbeddings generates embeddings for a single text
func (v *VoyageAIService) GenerateTextEmbeddings(text string) ([]float64, error) {
	input := Input{
		Content: []ContentItem{
			{Type: "text", Text: text},
		},
	}

	resp, err := v.callAPI([]Input{input}, InputTypeNull)
	if err != nil {
		return nil, err
	}

	if len(resp.Data) == 0 {
		return nil, fmt.Errorf("no embeddings returned")
	}

	return resp.Data[0].Embedding, nil
}

// GenerateImageEmbeddings generates embeddings for a single image
// Automatically compresses image if needed to meet Voyage AI requirements
func (v *VoyageAIService) GenerateImageEmbeddings(imgBytes []byte) ([]float64, error) {
	// Log original image info
	imgInfo, err := getImageInfo(imgBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to get image info: %w", err)
	}
	slog.Info("Original image",
		"width", imgInfo.Width,
		"height", imgInfo.Height,
		"pixels", imgInfo.Pixels,
		"size_mb", fmt.Sprintf("%.2f", imgInfo.SizeMB),
	)

	// Check if compression is needed
	isValid, errMsg := validateImageForVoyage(imgBytes)
	if !isValid {
		slog.Warn("Image validation failed, compressing...", "error", errMsg)
		imgBytes, err = compressImage(imgBytes, targetMaxPixels, defaultQuality)
		if err != nil {
			return nil, fmt.Errorf("failed to compress image: %w", err)
		}

		// Validate after compression
		isValid, errMsg = validateImageForVoyage(imgBytes)
		if !isValid {
			return nil, fmt.Errorf("image still invalid after compression: %s", errMsg)
		}
		slog.Info("Image compressed successfully and validated")
	} else {
		slog.Info("Image meets requirements, no compression needed")
	}

	// Convert to base64
	base64Img := imageToBase64(imgBytes)

	input := Input{
		Content: []ContentItem{
			{Type: "image_base64", ImageBase64: base64Img},
		},
	}

	resp, err := v.callAPI([]Input{input}, InputTypeNull)
	if err != nil {
		return nil, err
	}

	if len(resp.Data) == 0 {
		return nil, fmt.Errorf("no embeddings returned")
	}

	return resp.Data[0].Embedding, nil
}

// GenerateBatchEmbeddings generates embeddings for multiple inputs (text and/or images) in one API call
// Automatically compresses images if needed
func (v *VoyageAIService) GenerateBatchEmbeddings(inputs []interface{}) ([][]float64, error) {
	preparedInputs := make([]Input, 0, len(inputs))

	for idx, item := range inputs {
		switch val := item.(type) {
		case string:
			// Text input
			preparedInputs = append(preparedInputs, Input{
				Content: []ContentItem{
					{Type: "text", Text: val},
				},
			})

		case []byte:
			// Image bytes - compress if needed
			imgBytes := val

			imgInfo, err := getImageInfo(imgBytes)
			if err != nil {
				return nil, fmt.Errorf("input %d: failed to get image info: %w", idx, err)
			}
			slog.Info(fmt.Sprintf("Input %d: Image", idx),
				"width", imgInfo.Width,
				"height", imgInfo.Height,
				"pixels", imgInfo.Pixels,
				"size_mb", fmt.Sprintf("%.2f", imgInfo.SizeMB),
			)

			isValid, errMsg := validateImageForVoyage(imgBytes)
			if !isValid {
				slog.Warn(fmt.Sprintf("Input %d: Compression needed", idx), "reason", errMsg)
				imgBytes, err = compressImage(imgBytes, targetMaxPixels, defaultQuality)
				if err != nil {
					return nil, fmt.Errorf("input %d: failed to compress image: %w", idx, err)
				}

				// Validate after compression
				isValid, errMsg = validateImageForVoyage(imgBytes)
				if !isValid {
					return nil, fmt.Errorf("input %d: still invalid after compression: %s", idx, errMsg)
				}
				slog.Info(fmt.Sprintf("Input %d: Compressed successfully", idx))
			} else {
				slog.Info(fmt.Sprintf("Input %d: No compression needed", idx))
			}

			base64Img := imageToBase64(imgBytes)
			preparedInputs = append(preparedInputs, Input{
				Content: []ContentItem{
					{Type: "image_base64", ImageBase64: base64Img},
				},
			})

		default:
			return nil, fmt.Errorf("unsupported input type at index %d: %T", idx, item)
		}
	}

	// Single API call for all inputs
	slog.Info("Generating embeddings", "count", len(preparedInputs))
	resp, err := v.callAPI(preparedInputs, InputTypeNull)
	if err != nil {
		return nil, err
	}
	slog.Info("Successfully generated embeddings", "count", len(resp.Data))

	// Extract embeddings in order
	embeddings := make([][]float64, len(resp.Data))
	for _, data := range resp.Data {
		embeddings[data.Index] = data.Embedding
	}

	return embeddings, nil
}

// GenerateEmbeddingsWithType generates embeddings with a specific input type (query/document)
func (v *VoyageAIService) GenerateEmbeddingsWithType(text string, inputType InputType) ([]float64, error) {
	input := Input{
		Content: []ContentItem{
			{Type: "text", Text: text},
		},
	}

	resp, err := v.callAPI([]Input{input}, inputType)
	if err != nil {
		return nil, err
	}

	if len(resp.Data) == 0 {
		return nil, fmt.Errorf("no embeddings returned")
	}

	return resp.Data[0].Embedding, nil
}

// GenerateMultimodalEmbeddings generates embeddings for content that combines text and image
func (v *VoyageAIService) GenerateMultimodalEmbeddings(text string, imgBytes []byte) ([]float64, error) {
	// Compress image if needed
	isValid, errMsg := validateImageForVoyage(imgBytes)
	if !isValid {
		slog.Warn("Image validation failed, compressing...", "error", errMsg)
		var err error
		imgBytes, err = compressImage(imgBytes, targetMaxPixels, defaultQuality)
		if err != nil {
			return nil, fmt.Errorf("failed to compress image: %w", err)
		}
	}

	base64Img := imageToBase64(imgBytes)

	input := Input{
		Content: []ContentItem{
			{Type: "text", Text: text},
			{Type: "image_base64", ImageBase64: base64Img},
		},
	}

	resp, err := v.callAPI([]Input{input}, InputTypeNull)
	if err != nil {
		return nil, err
	}

	if len(resp.Data) == 0 {
		return nil, fmt.Errorf("no embeddings returned")
	}

	return resp.Data[0].Embedding, nil
}

// callAPI makes the actual API request to VoyageAI
func (v *VoyageAIService) callAPI(inputs []Input, inputType InputType) (*EmbeddingResponse, error) {
	reqBody := EmbeddingRequest{
		Inputs:    inputs,
		Model:     v.model,
		InputType: inputType,
	}

	jsonBody, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("POST", baseURL, bytes.NewReader(jsonBody))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+v.apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := v.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		var apiErr APIError
		if json.Unmarshal(body, &apiErr) == nil && apiErr.Error.Message != "" {
			return nil, fmt.Errorf("API error (%d): %s", resp.StatusCode, apiErr.Error.Message)
		}
		return nil, fmt.Errorf("API error (%d): %s", resp.StatusCode, string(body))
	}

	var embeddingResp EmbeddingResponse
	if err := json.Unmarshal(body, &embeddingResp); err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %w", err)
	}

	return &embeddingResp, nil
}
